Feature: Shopping List

rueblimaster opened the pull request:
## Dependencies

- #5244

## What
Adds the base frame for the general Shopping List. Only way to interact with it for now is by the `/shshoppinglist` command

### What is coming:

* Categories
* Persistent storage
* Recipes
* Buttons to add stuff to the list

- #4640 
- #4656 

<details>
<summary>Images</summary>

<!-- drop images here -->
<img width="257" height="101" alt="image" src="https://github.com/user-attachments/assets/1845fb96-6b18-4061-bf11-a67947aed07c" />

</details>

## Changelog New Features
+ Added generic Shopping List. - rueblimaster
    * Use via `/shshoppinglist`.

rueblimaster committed:
Create ShoppingList.kt

rueblimaster committed:
Create RecipeInventory.kt

rueblimaster committed:
Create RecipeResolver.kt

rueblimaster committed:
made primitive recipe storage capable

rueblimaster committed:
Merge branch 'beta' into shopping-list-rework

rueblimaster committed:
Revert "made primitive recipe storage capable"

This reverts commit 1b12f7f125698373c2fc67fb447b130f3443c1f1.

rueblimaster committed:
Revert "Fix: Some config stuff that June found (#4628)"

This reverts commit 4e790d301669f7d5e15e049b43b60850b1437d44.

rueblimaster committed:
working commands

rueblimaster committed:
item logic

rueblimaster committed:
chat messages upon commands

rueblimaster committed:
Add quotes to item names in chat messages

Chat messages in ShoppingList now display item names in single quotes for improved clarity when adding, removing, or updating items.

rueblimaster committed:
add config and implement basic rendering

rueblimaster committed:
Reapply "Fix: Some config stuff that June found (#4628)"

This reverts commit 7ba6e254ce00de18f4e6a6b7bdea7b95462e4478.

rueblimaster committed:
removed unused recipe stuff

rueblimaster committed:
remove wrong TODO

Thunderblade73 commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingList.kt:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    private fun buildDisplay() {
+        display = buildList {
+            items.forEach { addAll(it.buildDisplay()) }
+        }
```
```suggestion
        display = items.flatMap(ShoppingListItem::buildDisplay)
```

Thunderblade73 commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingList.kt:
```diff
@@ -0,0 +1,116 @@
+object ShoppingList {
+    val config = SkyHanniMod.feature.inventory.shoppingList
+
+    val items = mutableListOf<ShoppingListItem>()
```
Make this a ``mutableMapOf<NeuInternalName,ShoppingListItem>()`` as you request the internalName often (which with the current implementation runs in O(n) but the map runs in O(1) )

rueblimaster committed:
Refactor ShoppingList items to use a map

Changed ShoppingList items from a mutable list to a mutable map keyed by NeuInternalName. This improves lookup efficiency and simplifies add/remove operations for shopping list items.

rueblimaster committed:
Refactor buildDisplay to use flatMap

Simplified the buildDisplay function by replacing manual list building with flatMap on items' values. This improves readability and conciseness.

rueblimaster committed:
Revert "Reapply "Fix: Some config stuff that June found (#4628)""

This reverts commit 222b2e4983500411b9ffe37c435f26d741eb36fb.

Thunderblade73 commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingList.kt:
```diff
@@ -0,0 +1,115 @@
+        buildDisplay()
+    }
+
+    private fun getItemOrNull(internalName: NeuInternalName) = items.getOrDefault(internalName, null)
```
``items[internalName]`` is the same thing. So first of all replace the getOrDefault and then inline this function.

rueblimaster committed:
Merge branch 'beta' into shopping-list-base

rueblimaster committed:
Reapply "Reapply "Fix: Some config stuff that June found (#4628)""

This reverts commit 3af0e464d83cebb9c243776e6a608baabd4b204e.

rueblimaster committed:
Refactor item lookup in ShoppingList methods

Replaced usage of getItemOrNull with direct access to items[internalName] in add and remove methods for improved clarity and consistency.

rueblimaster committed:
Merge branch 'beta' into shopping-list-base

rueblimaster committed:
Merge branch 'beta' into shopping-list-base

rueblimaster committed:
Merge branch 'beta' into shopping-list-base

rueblimaster committed:
fix parseItemNameTabComplete

rueblimaster committed:
switch from internalName to itemName InternalNameArgumentType

rueblimaster committed:
switch to double in amounts

rueblimaster committed:
Merge branch 'beta' into shopping-list-base

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingListItem.kt:
```diff
@@ -0,0 +1,22 @@
+    @Expose
+    val internalName: NeuInternalName,
+    @Expose
+    var amount: Double,
```
why is this a double?

rueblimaster replied:
because of item stacks and coins

hannibal002 replied:
well, item stacks is no reason to have double :P

but coin is fair, can you add a comment about this in the code, please

rueblimaster replied:
sure

hannibal002 reviewed (changes requested):
Thanks for splitting the larger feature into smaller PRs, that makes reviewing much easier! A few things came up:

Context in the PR description: Since this is the base frame of a larger feature, it would really help reviewers to mention that in the description and link the follow-up PRs. Right now, it's hard to judge architectural decisions without knowing what's coming next.

BrigadierUtils change: The .map { it.replace(" ", "_") } change looks unrelated to the Shopping List. Could you move that into its own PR?

Config toggle: What does the enabled toggle control here? The feature is entirely command-driven, so it's unclear what disabling it would do. If it controls the overlay rendering, that makes sense, but then the command would still work while the toggle is off, which could be confusing. Worth clarifying what the intended UX is.

Persistence: The items map lives only in memory right now, so the list is lost on restart. Is that intentional for this base PR, with persistence coming in a follow-up? If so, mentioning that in the description would help.

Overall the structure looks clean and follows project patterns well (command registration, RenderDisplayHelper, config accordion). Just needs some clarification on the points above before merging.

rueblimaster committed:
Merge branch 'beta' into shopping-list-base

rueblimaster committed:
add clarifying comment

rueblimaster committed:
make config toggle make sense

rueblimaster committed:
Merge remote-tracking branch 'upstream/beta' into shopping-list-base

rueblimaster committed:
revert unneeded stuff

rueblimaster commented:
@hannibal002 
I addressed all your points

hannibal002 reviewed (approved):
code looks good for me.

please expand the changelog wording a bit to explain whats the point of this command, why users might wanna use this.
